engine/runtime-integration-tests/src/test/java/org/enso/interpreter/test/ValuesGenerator.java

```diff
@@ -369,6 +369,10 @@ public List<Value> numbers() {
     if (languages.contains(Language.ENSO)) {
       collect.add(v(null, "", "42").type());
       collect.add(v(null, "", "6.7").type());
+      collect.add(v(null, "", "9223372036854776000").type());
+      collect.add(v(null, "", "9223372036854775999").type());
+      collect.add(v(null, "", "9223372036854776000.0").type());
+      collect.add(v(null, "", "9223372036854775999.0").type());
```

Comment:
@GregoryTravis, please add here "all" the Enso numbers that you consider suspicious.

Reply:
On this branch, the only value that has ==/hash inconsistency is 9223372036854775808, which I have added.

Reply:
When I run the tests on this branch:

```
sbt:enso> runtime-integration-tests/testOnly *HashCodeTest
[...]
[info] Passed: Total 0, Failed 0, Errors 0, Passed 0
[info] No tests to run for runtime-integration-tests / Test / testOnly
```

```
sbt:enso> project runtime-integration-tests
[info] set current project to runtime-integration-tests (in build file:/Users/gmt/dev/enso/enso/)
sbt:runtime-integration-tests> runtime-integration-tests/testOnly *HashCodeTest*
[...]
[info] Passed: Total 0, Failed 0, Errors 0, Passed 0
[info] No tests to run for Test / testOnly
```

Reply:
I am not sure what can be wrong? `sbt buildEngineDistribution` first, maybe?

Reply:
> On this branch, the only value that has ==/hash inconsistency is 9223372036854775808, which I have added.

Great example! `9223372036854775808 == 2^63` `BigDecimal` for both `double` and `BigInteger` are equal, but the hashcodes aren't.

Reply:
I believe the `2^63` value if fixed by 0413d5f and 96da44b - the other commit uses `new BigDecimal(double)` instead of `valueOf` - they behave quite differently! Looks like the `new BigDecimal()` approach is the right one for the `2^63` case. We should review also other usages of `BigDecimal.valueOf`:

![Usage of BigDecimal.valueOf](https://github.com/enso-org/enso/assets/26887752/68d1a5a1-ebd6-493c-be56-59e10b7baab1)

Reply:
@JaroslavTulach This is failing in `0.1.truncate`, since it uses `.toBigIntegerExact` in TruncateNode, and that throws an exception when the argument is not an integer.

Reply:
Also, getting an inconsistent hash for `Float.max_value`:

```
    d = Decimal.new Float.max_value
    IO.println (d == Float.max_value)
```

```
Execution finished with an error: Assertion Error: 'java.lang.AssertionError: Different hash code! Original 1.7976931348623157E308[#318db821] got converted to WithWarnings{Value JavaObject[1.7976931348623157E
```

Reply:
I can see this is getting pretty complicated. To support `Decimal.from (that : Float)` conversion, we need to make sure the `hash` of original and new value is the same. That leads to `new BigInteger(that)`. However then `(double)0.1` is converted to a nasty number with fifteen+ decimal digits.

On the other hand, use of `BigInteger.valueOf` may lead to `Decimal` objects with different `hash` than the original one when doing `Decimal.from(that:Float)` conversion. My attempt to deal with it is visible in TBD - let's see how well it goes with the tests.

Reply:
I am curious, because this hash-caching technique might be a problem if we create the same `Decimal` value from a string, without the benefit of a `Number` to get a hash value from. This might allow two Decimals of the same value, one with a cached hash and one without. These values would be `==` but have different hashes.

I do like the idea you mentioned of converting everything to a double, or something like that -- that might make this unnecessary.

Reply:
> What was the test case that was fixed by [f397947](https://github.com/enso-org/enso/commit/f39794796c7f185c732fc08247d187282a58a24d)?

That fixes various [AssertionError](https://github.com/enso-org/enso/actions/runs/8749250691/job/24010598281?pr=9740#step:7:1209)s - if there is a `from` conversion, then the newly created atom must have the same hashcode as the old value.

> I am curious, because this hash-caching technique might be a problem if we create the same `Decimal` value from a string, without the benefit of a `Number` to get a hash value from. 

Yes, those would be different values when hashed and they should technically _not be equal_.

> This might allow two Decimals of the same value, one with a cached hash and one without. These values would be `==` but have different hashes.

Alas, that is true...

> I do like the idea you mentioned of converting everything to a double, or something like that -- that might make this unnecessary.

...OK, I'll try it. It seems like the only way out of the current misery.

Reply:
Here is my attempt to use `Double.hashCode` for hashing: https://github.com/enso-org/enso/pull/9740/commits/867262faedf6c7fa4fa6c5b066025a7b9506a526 It requires the `Decimal.from (that : Float)` **conversion to be exact** - e.g. `new BigDecimal(value)`- without that we'd get various [AssertionError](https://github.com/enso-org/enso/actions/runs/8749250691/job/24010598281?pr=9740#step:7:1209)s back. These are the currently failing tests (mostly result of the exact conversion, I think):
```
[FAILED] min/max: [2/3, 320ms]
    - should calculate min and max correctly [141ms]
    - [FAILED] should calculate min and max correctly, with mixed Decimal and Int/Float [172ms]
        Reason: 12.1 did not equal 12.1 (at test/Base_Tests/src/Data/Decimal_Spec.enso:929:13-79).
    - should calculate min and max correctly, with Math.min/max [7ms]
--
[FAILED] comparison: [3/4, 335ms]
    - should compare correctly [185ms]
    - should compare correctly, even with different internal representations [44ms]
    - [FAILED] should compare correctly to Integer and Float [94ms]
        Reason: (Decimal.Value 0.1) did not equal 0.1 (at test/Base_Tests/src/Data/Decimal_Spec.enso:307:17-41).
    - should not implicitly convert Decimal to Integer in comparison [11ms]
[FAILED] arithmetic: [6/7, 278ms]
    - should allow arithmetic with Decimals, without Math_Context [113ms]
    - should allow arithmetic with Decimals, with precision setting (examples) [15ms]
--
    - [FAILED] should negate values correctly [16ms]
        Reason: (Decimal.Value -5.3) did not equal -5.3 (at test/Base_Tests/src/Data/Decimal_Spec.enso:510:13-56).
read_text: [6/6, 93ms]
--
[FAILED] text conversion: [3/4, 20ms]
    - should convert to text correctly [4ms]
    - should convert to text without scientific notation correctly [2ms]
--
    - [FAILED] should parse correctly [3ms]
        Reason: (Decimal.Value 123456789.87654) did not equal 1.2345678987654E8 (at test/Base_Tests/src/Data/Decimal_Spec.enso:717:13-74).
should correctly evaluate documentation examples: [8/8, 43ms]
--
[FAILED] Asserts: [4/5, 6ms]
    - [FAILED] should be enabled in tests [3ms]
        Reason: Expected a value of type Standard.Base.Errors.Common.Assertion_Error but got a value of type Standard.Base.Nothing.Nothing instead (at test/Base_Tests/src/Runtime/Asserts_Spec.enso:16:9-52).
    - should be able to take foreign functions as expressions [0ms]
--
[FAILED] conversions: [7/8, 215ms]
    - should convert correctly to and from Integer [35ms]
    - should convert correctly to Float [37ms]
--
    - [FAILED] Decimal.to_float should compare correctly with the original Decimal [7ms]
        Reason: Expected False to be True (at test/Base_Tests/src/Data/Decimal_Spec.enso:594:13-53).
    - Decimal.to_integer should compare correctly with the original Decimal [4ms]
--
[FAILED] construction: [17/19, 135ms]
    - should be able to construct a Decimal from a string [2ms]
    - should throw Number_Parse_Error on a badly-formatted string [4ms]
--
    - [FAILED] should be able to explicity and implicitly specify infinite precision [5ms]
        Reason: (Decimal.Value 123.13) did not equal 123.13 (at test/Base_Tests/src/Data/Decimal_Spec.enso:83:13-76).
    - should report precision loss accurately [56ms]
--
    - [FAILED] should be convertible via .from [7ms]
        Reason: (Decimal.Value 123.4500000000000028421709430404007434844970703125) did not equal (Decimal.Value 123.45) (at test/Base_Tests/src/Data/Decimal_Spec.enso:120:13-67).
    - constructor should respect Math_Context  [7ms]
--
[FAILED] signs: [1/3, 7ms]
    - [FAILED] should calculate abs correctly [3ms]
        Reason: (Decimal.Value 12.345) did not equal 12.345 (at test/Base_Tests/src/Data/Decimal_Spec.enso:728:13-60).
    - [FAILED] should calculate negate correctly [1ms]
        Reason: (Decimal.Value -12.345) did not equal -12.345 (at test/Base_Tests/src/Data/Decimal_Spec.enso:734:13-64).
    - should calculate signum correctly [1ms]
--
[FAILED] remainder: [2/3, 40ms]
    - [FAILED] should define remainder [37ms]
        Reason: (Decimal.Value 0.2) did not equal 0.2 (at test/Base_Tests/src/Data/Decimal_Spec.enso:624:17-67).
    - should define remainder (example) [0ms]
--
[FAILED] rounding: [16/22, 162ms]
    - Can round positive decimals correctly [27ms]
    - Can round negative decimals correctly [13ms]
--
    - [FAILED] Can round positive decimals to a specified number of decimal places [4ms]
        Reason: (Decimal.Value 3.14) did not equal 3.14 (at test/Base_Tests/src/Data/Round_Spec.enso:42:9-46).
    - [FAILED] Can round negative decimals to a specified number of decimal places [4ms]
        Reason: (Decimal.Value -3.14) did not equal -3.14 (at test/Base_Tests/src/Data/Round_Spec.enso:55:9-48).
    - Can round positive decimals to a specified negative number of decimal places [17ms]
--
    - [FAILED] Banker's rounding handles half-way values correctly [13ms]
        Reason: (Decimal.Value 0.24) did not equal 0.24 (at test/Base_Tests/src/Data/Round_Spec.enso:106:9-62).
    - Banker's rounding handles non-half-way values just like normal rounding [19ms]
    - [FAILED] Can round correctly near the precision limit [4ms]
        Reason: (Decimal.Value 1.2222222223) did not equal 1.2222222223 (at test/Base_Tests/src/Data/Round_Spec.enso:132:9-62).
    - [FAILED] Can round correctly near the precision limit, using banker's rounding [5ms]
        Reason: (Decimal.Value 1.2222222222) did not equal 1.2222222222 (at test/Base_Tests/src/Data/Round_Spec.enso:157:9-79).
    - [FAILED] Floating point imperfect representation counter-examples [4ms]
        Reason: (Decimal.Value 1.22) did not equal 1.22 (at test/Base_Tests/src/Data/Round_Spec.enso:182:9-62).
    - Can round small integers to a specified number of decimal places correctly (value is unchanged) [2ms]
```

Reply:
Maybe we should [attach a warning](https://github.com/enso-org/enso/pull/9740/commits/9619afe8d16a4aaeb3694144fcc754eb8ed1e54c) doing no-parameter `Decimal.from (that : Float)` that the conversion may be _too precise_!?

Reply:
> Maybe we should [attach a warning](https://github.com/enso-org/enso/pull/9740/commits/9619afe8d16a4aaeb3694144fcc754eb8ed1e54c) doing no-parameter `Decimal.from (that : Float)` conversion. The conversion may be _too precise_ for regular user expectation!?

That sounds like a good idea

Reply:
At least some of these test failures can be fixed by changing them from

```
            12.1 . min (Decimal.new "12.3") . should_equal (Decimal.new "12.1")
```
to
```
            12.1 . min (Decimal.new "12.3") . should_equal (Decimal.new 12.1)
```
(And using exact conversions everywhere.)

This would mean that `Decimal.new 12.1 != Decimal.new "12.1")`, which is strange, but is also what serious users of Java Decimal expect -- it is documented that the string argument is the preferred one.

What is *not* clear is how this is going to work with the database tables.

I will try this out and see how it goes.

Reply:
As discussed on the standup, I think it may make sense to go for the _exact_ conversion by default (so that hashing will use it), but we can use the _'optional arguments of `.from`'_ to allow the user to also opt-in for the more familiar conversion.

Like
```
Decimal.from (that:Float) (exact:Boolean=True) = ...
```

Reply:
Right now (after [8f475d1](https://github.com/enso-org/enso/pull/9740/commits/8f475d1fbc0587d7941711baf3d8df83022ab51d)) it is:
- `Decimal.from 3.14` - exact precision
- `Decimal.new 3.14` - non-exact precision

e.g. it is _not necessary_ to add the `exact` argument to `from` conversion as we can choose the precision by calling either `from` or `new`.